src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+	/* 0x80 - SINOWEALTH_XY_INDEPENDENT */
+	uint8_t dpi_count:4;
+	uint8_t active_dpi:4;
+	uint8_t dpi_enabled;
+	/* bit set: disabled, unset: enabled
+	 * this structure has support for eight DPI slots,
+	 * but the glorious software only exposes six
+	 */
+	uint8_t dpi[16];
+	/* DPI/CPI is encoded in the way the PMW3360 sensor accepts it
+	 * value = (DPI - 100) / 100
+	 * If XY are identical, dpi[0-6] contain the sensitivities,
+	 * while in XY independent mode each entry takes two chars for X and Y.
+	 */
+	struct RGB8 dpi_color[8];
+	uint8_t rgb_effect;
+	/* see enum rgb_effect */
+	uint8_t glorious_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t glorious_direction;
+	uint8_t single_mode;
+	struct RBG8 single_color;
+	uint8_t breathing_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t breathing_colorcount;
+	/* 7, constant */
+	struct RBG8 breathing_colors[7];
+	uint8_t tail_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t unk4[33];
+	uint8_t rave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	struct RBG8 rave_colors[2];
+	uint8_t wave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t breathing1_mode;
+	/* 0x1/2/3 - speed */
+	struct RBG8 breathing1_color;
+	uint8_t unk5;
+	uint8_t lift_off_distance;
+	/* 0x1 - 2 mm
+	 * 0x2 - 3 mm
+	 */
+	uint8_t padding[520 - 131];
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct sinowealth_config_report) == SINOWEALTH_CONFIG_SIZE, "Invalid size");
+
+struct sinowealth_data {
+	/* this is kinda unnecessary at this time, but all the other drivers do it too ;) */
+	struct sinowealth_config_report config;
+};
+
+static int
+sinowealth_raw_to_dpi(int raw)
+{
+	return (raw + 1) * 100;
+}
+
+static int
+sinowealth_dpi_to_raw(int dpi)
+{
+	return dpi / 100 - 1;
+}
+
+static struct ratbag_color
+sinowealth_raw_to_color(struct RGB8 raw)
+{
+	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+}
+
+static struct RGB8
+sinowealth_color_to_raw(struct ratbag_color color)
+{
+	return (struct RGB8) {.r = color.red, .g = color.green, .b = color.blue};
+}
+
+static struct ratbag_color
+sinowealth_rbg_to_color(struct RBG8 raw)
+{
+	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+}
+
+static struct RBG8
+sinowealth_color_to_rbg(struct ratbag_color color)
+{
+	return (struct RBG8) {.r = color.red, .g = color.green, .b = color.blue};
+}
+
+static int
+sinowealth_rgb_mode_to_brightness(uint8_t mode)
+{
+	return max(mode >> 4 << 6, 255);
+}
+
+static int
+sinowealth_rgb_mode_to_speed(uint8_t mode)
+{
+	switch (mode & SINOWEALTH_RGB_SPEED_BITS) {
+	case 0: return 10000; /* static: does not translate to duration */
+	case 1: return 1500;
+	case 2: return 1000;
+	case 3: return 500;
+	default: return 0;
+	}
+}
+
+static uint8_t
+sinowealth_to_rgb_mode(int brightness, int speed_ms)
+{
+	uint8_t mode = (brightness + 1) >> 6 << 4;
```

Comment:
you should have the inverse helper to `sinowealth_rgb_mode_to_brightness` for this instead of open-coding it here, nicer for symmetry.

Reply:
I don't really agree on this one, because the usage in the code isn't symmetric, the helpers are asymmetric as well (bundling/unbundling multiple fields into a single field). I suppose if these really need to be symmetric they could be defined in terms of `struct ratbag_led`.

Reply:
Like so:

```c
static int
sinowealth_rgb_mode_to_brightness(uint8_t mode)
{
	/* brightness is 0-4 in the upper nibble of mode,
	 * while libratbag uses 0-255.
	 */
	return max((mode >> 4) * 64, 255);
}

static int
sinowealth_rgb_mode_to_speed(uint8_t mode)
{
	switch (mode & SINOWEALTH_RGB_SPEED_BITS) {
	case 0: return 10000; /* static: does not translate to duration */
	case 1: return 1500;
	case 2: return 1000;
	case 3: return 500;
	default: return 0;
	}
}

static void
sinowealth_set_led_from_rgb_mode(const struct ratbag_led *led, uint8_t mode)
{
	led->brightness = sinowealth_rgb_mode_to_brightness(mode);
	led->ms = sinowealth_rgb_mode_to_speed(mode);
}

static uint8_t
sinowealth_led_to_rgb_mode(const struct ratbag_led *led)
{
	/* convert 0-255 to 0-4 in the upper nibble of mode */
	uint8_t mode = ((led->brightness + 1) / 64) << 4;
	if (led->ms <= 500) {
		mode |= 3;
	} else if (led->ms <= 1000) {
		mode |= 2;
	} else {
		mode |= 1;
	}
	return mode;
}
```